api/agent/agent.go

```diff
@@ -737,114 +745,183 @@ func newSizerRespWriter(max uint64, rw http.ResponseWriter) http.ResponseWriter
-	defer a.evictor.DeleteEvictToken(evictor)
 
 	statsUtilization(ctx, a.resources.GetUtilization())
+	state.UpdateState(ctx, ContainerStateStart, call.slots)
+
+	// stack unwind spelled out with strict ordering below.
 	defer func() {
-		statsUtilization(ctx, a.resources.GetUtilization())
-	}()
+		// IMPORTANT: we do not emit eviction errors to slot queue. Remember evictions occur
+		// in cases of overcapacity. If an accidental eviction occurs, then the client
+		// will have to wait hot poll period to spawn a new container.
+		if !evictor.isEvicted() {
+			select {
+			case <-initialized:
+			default:
+				tryQueueErr(models.ErrContainerInitFail, errQueue)
+			}
+
+			select {
+			case err := <-errQueue:
+				call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+			default:
+			}
+		}
 
-	defer tok.Close() // IMPORTANT: this MUST get called
+		if runRes != nil && runRes.Error() != context.Canceled {
+			logger.WithError(runRes.Error()).Info("hot function terminated")
+		}
 
-	state.UpdateState(ctx, ContainerStateStart, call.slots)
-	defer state.UpdateState(ctx, ContainerStateDone, call.slots)
+		// shutdown the container and related I/O operations and go routines
+		cancel()
 
-	container, err := newHotContainer(ctx, call, &a.cfg)
-	if err != nil {
-		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
-		return
-	}
-	defer container.Close()
+		// IMPORTANT: for release cookie (remove container), make sure ctx below has no timeout.
+		if cookie != nil {
+			cookie.Close(common.BackgroundContext(ctx))
+		}
 
-	logger := logrus.WithFields(logrus.Fields{"id": container.id, "app_id": call.AppID, "fn_id": call.FnID, "image": call.Image, "memory": call.Memory, "cpus": call.CPUs, "idle_timeout": call.IdleTimeout})
-	ctx = common.WithLogger(ctx, logger)
+		if container != nil {
+			container.Close()
+		}
 
-	ctx, cancel := context.WithTimeout(common.BackgroundContext(ctx), a.cfg.HotStartTimeout)
-	defer cancel()
+		lastState := state.GetState()
+		state.UpdateState(ctx, ContainerStateDone, call.slots)
 
-	cookie, err := a.driver.CreateCookie(ctx, container)
-	if err != nil {
-		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+		tok.Close() // release cpu/mem
+
+		// IMPORTANT: evict token is deleted *after* resource token.
+		// This ordering allows resource token to be freed first, which means once evict token
+		// is deleted, eviction is considered to be completed.
+		a.evictor.DeleteEvictToken(evictor)
+
+		statsUtilization(ctx, a.resources.GetUtilization())
+		if evictor.isEvicted() {
+			logger.Debugf("Hot function evicted")
+			statsContainerEvicted(ctx, lastState)
+		}
+	}()
+
+	go func() {
+		isInitialized, err := a.waitInitialize(ctx, caller, evictor, udsWait)
+		if err != nil {
+			tryQueueErr(err, errQueue)
+		}
+		if isInitialized {
+			close(initialized)
+		} else {
+			cancel()
+		}
+	}()
```

Comment:
initialized seems like largely a dupe of udsWait. this thread forwards the waiting on udsWait to receive to initialized. i think waitInitialize needs to exist to wait on any eviction events but the addition of udsWait into that seems only to happen so that we can turn the send of 1 error into udsWait into a 'closed channel' that can be received on multiple times, at least it seems slightly unnecessary. am not sure it's exactly the constraint, trying to deduce, anyway it'd be possible to get a similar effect with something like https://play.golang.org/p/SSZLKc-PxwN - where udsWait wouldn't return anything (ie hang on channel receive) if the container never ended up getting prepared for any other reason or the uds itself not being ready, but after that we could definitely say 'the container was initialized' to eg avoid queueing an error that initialization failed. just trying to simplify here if possible.

Reply:
yeah, it's left over from previous logic, etc. With timeouts in nanny go-routine this made more sense. It's gone now.